db/db_creator.py

```diff
@@ -1,70 +1,73 @@
 import logging, os
-from typing import Dict, Sequence
+from typing import Dict, List
 from urllib.parse import quote_plus
 
 # third-party libraries
-from sqlalchemy.engine import create_engine
+from sqlalchemy.engine import create_engine, Engine
 from yoyo import get_backend, read_migrations
 
+
+# Initialize settings and global variables
+
 logger = logging.getLogger(__name__)
 
-ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
-MIGRATIONS_PATH = os.path.join(ROOT_DIR, 'migrations')
+PARENT_PATH = os.path.dirname(os.path.abspath(__file__))
+MIGRATIONS_PATH = os.path.join(PARENT_PATH, 'migrations')
+
+# The metadata tables are append tables by default.
+DEFAULT_APPEND_TABLE_NAMES = ['job_run', 'data_source_status']
 
 
 class DBCreator:
 
     def __init__(
         self,
         db_params: Dict[str, str],
-        append_table_names: Sequence[str] = []
+        append_table_names: List[str] = []
     ) -> None:
 
-        self.db_name = db_params['dbname']
-        self.conn = None
-        self.conn_str = (
+        self.db_name: str = db_params['dbname']
+        self.conn_str: str = (
             'mysql+mysqldb' +
             f"://{db_params['user']}" +
             f":{quote_plus(db_params['password'])}" +
             f"@{db_params['host']}" +
             f":{db_params['port']}" +
             f"/{db_params['dbname']}?charset=utf8&ssl=true"
         )
-        self.engine = create_engine(self.conn_str)
-        self.append_table_names = append_table_names
-
-    def set_up(self) -> None:
-        logger.debug('set_up')
-        self.conn = self.engine.connect()
+        self.engine: Engine = create_engine(self.conn_str)
 
-    def tear_down(self) -> None:
-        logger.debug('tear_down')
-        self.conn.close()
+        self.append_table_names: List[str] = append_table_names
+        self.append_table_names += DEFAULT_APPEND_TABLE_NAMES
 
-    def get_table_names(self) -> Sequence[str]:
+    def get_table_names(self) -> List[str]:
         logger.debug('get_table_names')
         return self.engine.table_names()
 
-    def migrate(self) -> None:
+    def migrate(self) -> DBCreator:
         logger.debug('migrate')
         backend = get_backend(self.conn_str)
         migrations = read_migrations(MIGRATIONS_PATH)
         with backend.lock():
             backend.apply_migrations(backend.to_apply(migrations))
+        return self
 
-    def drop_records(self) -> None:
+    def drop_records(self) -> DBCreator:
         logger.debug('drop_records')
-        self.conn.execute('SET FOREIGN_KEY_CHECKS=0;')
+        conn = self.engine.connect()
+        conn.execute('SET FOREIGN_KEY_CHECKS=0;')
         for table_name in self.get_table_names():
             if 'yoyo' not in table_name and table_name not in self.append_table_names:
                 logger.debug(f'Table Name: {table_name}')
-                self.conn.execute(f'DELETE FROM {table_name};')
+                conn.execute(f'DELETE FROM {table_name};')
```

Comment:
Codacy is flagging this as a SQL Injection vulnerability, which is fair. Based on the research I've done, using parameters here isn't possible (https://stackoverflow.com/questions/40249590/sqlalchemy-error-when-adding-parameter-to-string-sql-query), and MyLA is doing [roughly the same thing in `cron.py`](https://github.com/tl-its-umich-edu/my-learning-analytics/blob/36a05c27eee83dd702d65666586ebebf0ab4cf0c/dashboard/cron.py#L78-L83).

I think the best way to fix this is to look into switching to an ORM, probably Django or SQLAlchemy, which I could look into as a separate issue.

Reply:
Aren't there modules that can check for and eliminate SQL injection?  It really doesn't seem like such a big deal, since the name comes from a query of the DB itself.

Reply:
Yeah, I don't think it's a big deal either right now. [mysqlclient, and by extension MySQLdb](https://mysqlclient.readthedocs.io/) might have a sanitizing function; I haven't looked there specifically. I think I saw somewhere that Postgres (`psycopg2`) has one.